[tagesschau] fix extraction

memo33 opened the pull request:
### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This fixes the extraction for tagesschau.de, the primary German news broadcaster, after an update of their website design in January.

Details:
- primarily fixes #28027
- the new website sends the video meta data as JSON encoded as an attribute of the player
- some old pages (e.g. [1]) still use the old format of `TagesschauIE`, so the old extraction code is kept as fallback
- the old player format is not used anymore, so `TagesschauPlayerIE` is removed
- several optional fields are added, in particular subtitles, resolving #21427
- crashes on empty playlists are fixed, resolving #13710
- a deprecation warning about inline regex flags is fixed in the old extraction code
- some test cases are updated, as some videos are not available anymore
- the video ID in one test case is changed from `ts-5727` to `video-45741` because it appears to be the more permanent identifier, which is also part of the player meta data; for example [1] is still available, while [2] cannot be accessed anymore
- the format identifiers are normalized to `s`,..,`xl` as previously used by `TagesschauPlayerIE` (instead of `webs.h264`,..,`webxl.h264`)

[1] https://www.tagesschau.de/multimedia/video/video-98529~_bab-sendung-209.html
[2] https://www.tagesschau.de/multimedia/sendung/bab/bab-3299~_bab-sendung-209.html

silsha commented:
Thank you for your work! I have made a few downloads and have not stumbled across any errors.

A little unfortunate is the `fulltitle` (and therefore the filename?) for media of type `broadcast`. In the case of the 20 Uhr Tagesschau, this is always "Ganze Sendung" - which is the `headline` of the first video file in the tagesschau JSON. More suitable here would be perhaps the `broadcastTitle` + the corresponding date.

Before everything broke the `fulltitle` in that case was e.g. `Broadcast: tagesschau \t21.01.2021 20:00 Uhr`.

Here are some examples of `broadcast` JSONs:

* https://www.tagesschau.de/api/multimedia/sendung/ts-41887.json
* https://www.tagesschau.de/api/multimedia/sendung/ts-41901.json
* https://www.tagesschau.de/api/multimedia/sendung/nm-5905.json

This is how the JSON file looked like in the old implementation: https://ia801809.us.archive.org/26/items/Tagesschau-2021-01-21/Sendung_-_tagesschau_21.01.2021_20_00_Uhr-ts-41115.info.json

memo33 commented:
Thank you for the feedback. Indeed, "Ganze Sendung" seems a bit too generic for the main evening broadcast. Unfortunately, the JSON included with the HTML pages is not the same as what you get with the API. I would like to avoid sending an additional request to the API just for the title, especially as it seems that it is being replaced by a new API at https://www.tagesschau.de/api2/, so I have added a workaround that attempts to take care of these important special cases by just using the data that is already available. Now one gets unique titles, for example for this playlist:

    $ python -m youtube_dl -e https://www.tagesschau.de/sendung/tagesschau/
    tagesschau 20 Uhr - 13.03.21 20:00
    tagesschau 20 Uhr - 12.03.21 20:00
    tagesschau 20 Uhr - 11.03.21 20:00
    tagesschau 20 Uhr - 10.03.21 20:00
    tagesschau 20 Uhr - 09.03.21 20:00
    tagesschau 20 Uhr - 08.03.21 20:00
    tagesschau 20 Uhr - 07.03.21 20:00

Additionally, I have removed duplicates from playlists, as the first entry tends to appear twice on the HTML pages.

silsha commented:
Thank you! That looks like a good solution. I'll try out a few more URLs tomorrow.

silsha commented:
By the way, I made a few more tests. Only the podcast[1] downloads don't seem to work - but I don't know if they did before. That would definitely be nice to have, but shouldn't block this PR.

[1] https://www.tagesschau.de/multimedia/podcasts/mal-angenommen-101~_episode-Keine-Nebenjobs-fuer-Politiker-Was-dann.html

memo33 commented:
Thanks for the thorough testing. I have had a look at the podcasts and it seems that extracting the URLs from them would have to be implemented quite differently, so it is probably best to leave that for another PR indeed, especially as this one is already quite large. Would you be able to provide code review?

silsha commented:
I can take a closer look at the code (and will later this week), but I am not a contributor to youtube-dl myself and can't add a code review to Github or speed up the PR in any way.

kl4ppdrachen commented:
I would ask if there is something new. I used it to stream in the morning the news on the TV via Home Assistant. 

I switched over to Spotify, but it would be nice, when it is working again.
